### PR TITLE
Flax sharded

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Adds support for FLAX sharding checkpoints

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Will merge `flax` before `tf` as the TF one still needs a few modification (mostly cleaning the documentation)

File: src/transformers/modeling_flax_utils.py [hunk 1]
```diff
@@ -788,10 +976,43 @@ def save_pretrained(self, save_directory: Union[str, os.PathLike], params=None,
...
-            params = params if params is not None else self.params
-            model_bytes = to_bytes(params)
-            f.write(model_bytes)
+
+        shards, index = flax_shard_checkpoint(params if params is not None else self.params, max_shard_size)
+        # Clean the folder from a previous save
+        for filename in os.listdir(save_directory):
+            full_filename = os.path.join(save_directory, filename)
+            # If we have a shard file that is not going to be replaced, we delete it, but only from the main process
+            # in distributed settings to avoid race conditions.
```
Comment (sanchit-gandhi) [inline src/transformers/modeling_flax_utils.py:985]: (nit) Main process / distributed settings are PyTorch specific
```suggestion
```

**Review by sanchit-gandhi — APPROVED** [review]
Very nice, thanks for adding this feature! Just a few small suggestions from me, but on the whole looks great!

File: src/transformers/modeling_flax_utils.py [hunk 2]
```diff
@@ -70,6 +78,88 @@ def quick_gelu(x):
...
+
+    sharded_state_dicts = []
+    current_block = {}
+    current_block_size = 0
+    total_size = 0
+
+    # flatten the weights to chunk
+    weights = flatten_dict(params, sep="/")
+    for item in weights:
+        weight_size = np.array(weights[item]).size * dtype_byte_size(weights[item].dtype)
```
Comment (patil-suraj) [inline src/transformers/modeling_flax_utils.py:129]: `weights` should already be a `np` or `jnp` array, why do we need to convert to `np.array` here ?

File: src/transformers/modeling_flax_utils.py [hunk 3]
```diff
@@ -333,6 +423,38 @@ def to_fp16(self, params: Union[Dict, FrozenDict], mask: Any = None):
...
+        """
+        Args:
+        This is the same as
+        [`flax.serialization.from_bytes`](https:
+            //flax.readthedocs.io/en/latest/_modules/flax/serialization.html#from_bytes)
+        but for a sharded checkpoint. This load is performed efficiently:
+            each checkpoint shard is loaded one by one in
+        RAM and deleted after being loaded in the model.
+            shard_files (`List[str]`:
+                The list of shard files to load.
```
Comment (patil-suraj) [inline src/transformers/modeling_flax_utils.py:437]: (nit) the doc formatting looks a bit weird here.

File: src/transformers/modeling_flax_utils.py [hunk 4]
```diff
@@ -498,6 +624,10 @@ def from_pretrained(
                 elif os.path.isfile(os.path.join(pretrained_model_name_or_path, FLAX_WEIGHTS_NAME)):
                     # Load from a Flax checkpoint
                     archive_file = os.path.join(pretrained_model_name_or_path, FLAX_WEIGHTS_NAME)
+                elif os.path.isfile(os.path.join(pretrained_model_name_or_path, FLAX_WEIGHTS_INDEX_NAME)):
+                    # Load from a sharded PyTorch checkpoint
```
Comment (patil-suraj) [inline src/transformers/modeling_flax_utils.py:628]:
```suggestion
                    # Load from a sharded Flax checkpoint
```

File: src/transformers/modeling_flax_utils.py [hunk 5]
```diff
@@ -518,9 +648,11 @@ def from_pretrained(
                     pretrained_model_name_or_path,
                     filename=filename,
                     revision=revision,
+                    # TODO raise an issue here as pytroch version uses a mirror argument
```
Comment (patil-suraj) [inline src/transformers/modeling_flax_utils.py:651]: is this `TODO` still standing ?

File: tests/test_modeling_flax_common.py [hunk 6]
```diff
@@ -991,6 +994,59 @@ def _assert_all_params_initialised(model, params):
...
+
+                index_file = os.path.join(tmp_dir, FLAX_WEIGHTS_INDEX_NAME)
+                # Check there is an index but no regular weight file
+                self.assertTrue(os.path.isfile(index_file))
+                self.assertFalse(os.path.isfile(os.path.join(tmp_dir, FLAX_WEIGHTS_NAME)))
+
+                # Check a file is bigger than max_size only when it has a single weight
+                for shard_file, size in shard_to_size.items():
+                    if max_size.endswith("kiB"):
+                        max_size_int = int(max_size[:-3]) * 2**10
```
Comment (patil-suraj) [inline tests/test_modeling_flax_common.py:1027]: Very nice test!

**Review by patil-suraj — APPROVED** [review]
Great job @ArthurZucker , looks good to me! Just left some nits.

**Review by sgugger — APPROVED** [review]
Thanks a lot for working on this!

File: src/transformers/modeling_flax_utils.py [hunk 7]
```diff
@@ -70,6 +78,88 @@ def quick_gelu(x):
 }
 
 
+def dtype_byte_size(dtype):
+    """
+    Returns the size (in bytes) occupied by one parameter of type `dtype`. Example:
+    ```py
+    >>> dtype_byte_size(torch.float32)
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/modeling_flax_utils.py:85]: Yes, doc should be `np.float32` I think :) Nice catch

File: src/transformers/modeling_flax_utils.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/modeling_flax_utils.py:437]: Reformatted it

**Pull request merged**